library(shiny)
library('shiny.semantic')
library('tidyverse')
library('vroom')
library('janitor')
library('DT')
library('plotly')

ui <- semanticPage(
  #theme = bslib::bs_theme(bootswatch = "litera"),
  sidebarLayout(
    sidebarPanel(
      img(src = "https://www.sandiegocounty.gov/content/sdc/hhsa/_jcr_content/iparside/sdcdeptlogo.img.jpg/1580776450704.jpg",
          width = "310px", height = "100px"),
      tags$h2("Overview"),
      tags$em("This application is made for internal use and is provided 'as-is'."),
      p("This is a quick application for data QC from the files generated by the Nephele pipeline from NIH. This is valuable until Thieagen/terra builds out coverage estimations from their pipeline."),
      tags$h2(""),
      tags$h2("Instructions"),
      p("Input as many of the 'outputs/metrics/[sample_name]_down_depth_out.txt' files as you like to get an understanding of genome coverage per primer pair. It should run pretty fast, but may take a few seconds to upload/display. Currently it is made for Artic v3 primers, but a radio button for switching to v4/v4.1 primer positions is on the agenda."),
      tags$a(href="https://nephele.niaid.nih.gov/", 
             "Nephele pipeline (link)."),
      tags$h2(""),
      tags$a(href="https://github.com/benkraj/", "App/code by Ben Krajacich, using Shiny/Shiny.semantic"),
      tags$h2("")
    ),
    main_panel(
      h2("File Input:"),
      fileInput("upload", NULL, buttonLabel = "Upload...", multiple = TRUE, accept = ".txt"),
      semantic_DTOutput("data.format"),
      h2("Depth of coverage per primer:"),
      plotlyOutput(outputId = "interactive.coverage")
      #plotOutput("coverage.plot")
      #dataTableOutput("data.format")
    )
  )
  ,
)
server <- function(input, output, session) {
  #output$files <- renderTable(input$upload)
  
  data <- reactive({
    req(input$upload)
    
    formatted.data <- 
      # map2_df(input$upload$name, input$upload$datapath, 
      #         ~fread(.y) %>% mutate(sample_id = .x)) %>%
       map2_df(input$upload$datapath,
              input$upload$name,
              ~vroom(.x, col_names = c("reference", "pos", "downsampled_depth")) %>%
                mutate(sample_id = gsub("_down_depth_out.txt","", basename(.y)))) %>%

      #select(sample_id=file, nc_045512_2, pos=x1, downsampled_depth=x0) %>%
      mutate(downsampled_depth = case_when(downsampled_depth >= 200 ~ 200, 
                                           TRUE ~ downsampled_depth),
             coverage = case_when(downsampled_depth == 0 ~ "0", 
                                  downsampled_depth > 0 & downsampled_depth < 10 ~ "0-9",
                                  downsampled_depth >= 10 & downsampled_depth < 20 ~ "10-19",
                                  downsampled_depth >= 20 & downsampled_depth < 50 ~ "20-49",
                                  downsampled_depth >= 50 & downsampled_depth < 100 ~ "50-99",
                                  downsampled_depth >= 100 ~ "100+"),
             coverage = factor(coverage, levels=c("0", "0-9", 
                                                  "10-19", "20-49", 
                                                  "50-99", "100+")),
             gene = case_when(pos >= 266 & pos <= 21555 ~ "ORF1ab",
                              pos >= 21563 & pos <= 25384 ~ "S",
                              pos >= 28274 & pos <= 29533 ~ "N",
                              TRUE ~ "Other")) 
  })
  
  output$data.format <- renderDataTable({
    semantic_DT(data() %>% 
                count(sample_id, coverage) %>% 
                ungroup() %>% 
                group_by(sample_id) %>% 
                mutate(percent = n/sum(n)) %>% 
                select(-n) %>%
                pivot_wider(., names_from="coverage", values_from="percent") 
              , 
              caption = htmltools::tags$caption(style = 'caption-side: top; text-align: center; color:black; font-size:150% ;','Percentage of genome covered at depth')) %>%
      DT::formatPercentage(columns=2:7, digits=2)
  })

  # output$coverage.plot <- renderPlot({
  #   ggplot(data()) +
  #     geom_line(aes(x=pos, y=downsampled_depth)) +
  #     geom_area(aes(x=pos, y=downsampled_depth), fill='darkblue', alpha=0.6) +
  #     scale_x_continuous(breaks=scales::pretty_breaks(n=28)) +
  #     theme(axis.text.x = element_text(angle = 90)) +
  #     scale_y_continuous(
  #       trans = "log10",
  #       breaks = c(1:10, seq(20,100,10), seq(200,1700,300))
  #     ) +
  #     ggtitle("Coverage Across SARS-CoV-2 Genome") +
  #     facet_wrap(~sample_id, ncol=3)
  # })
    
    output$interactive.coverage <- renderPlotly({
      source('primer.coverage.downsampled.R', local=T)
      
      primer.coverage.downsampled(data()) %>%
        group_by(sample_id) %>%
        do(p=plot_ly(., x = ~primername, y = ~mean_downsampled_depth, type = "bar", name=~sample_id) %>% 
             layout(yaxis= list(title = FALSE),
                    yaxis = list(tickfont = list(size = 3))) %>%
             layout(xaxis= list(title = "Primer"))) %>%
        subplot(nrows = round(length(unique(data()$sample_id))/2.5), shareX = TRUE, shareY = TRUE) }
    )
}


shinyApp(ui, server)
